### PR TITLE
Add gdi09ea

### PR DESCRIPTION
**Pull request by fruestueck**
Copy of #16834 (after squashing 400 bleed commits)

Adapted to global mission script

---
Add mission 9 "[Secure the Danube](https://cnc.fandom.com/wiki/Secure_the_Danube)" for the Tiberian Dawn GDI campaign (#4988).

The Lua script includes parts from [scg07ea](https://github.com/OpenRA/OpenRA/pull/11090) and [scb09ea](https://github.com/OpenRA/OpenRA/pull/11306).

Have fun testing! : )

~Original behavior:
```
+ Send one Gunboat (three in total) for each...
 * GDI Units near (CellTriggers) North NodBunker
 * First NodBunker is destroyed
 * Second NodBunker is destroyed

+ If a gunboat reaches its destination (West side of Map)
 * it will continue its patrol horizontally
 * despite the MissionGoal, the game is NOT over.

+ gunboats don't attack and get destroyed by one obelisk attack

+ A chinook is provided as soon as the south NodBunkes (gun3, gun4) get destroyed
 
+ Player WINs only when all Nod structures and units are destroyed.

+ Kill all four sam sites to gain airstrike support power

+ church does not spawn a crate (should we?)
```

Changes:
- Prerequesits for e4, ltnk & arty reduced to ~~techlevel.low~~ ~afld and barracks respectivly
- A `PatrolProductionQueue` and `ReplenishPatrolUnit()` function added for HandofNod.
OnDeath grd1 and grd2 Units will be rebuild and sent back to patrol
- For the Nod1 to 5 and Auto1 to 8 AttackWaves the lua-script (from mission scb09ea) is just grabbing available (`IsIdle`) Untis. As the Airfield and HandofNod produces these units by random, some attack waves may be incomplete at different playthrough.
This could be changed to spawn the required units at GameStart. Alternatively the required Units could be built with a queue similar to `ReplenishPatrolUnit()`. This would take much more effort though and would be a possible feature PR.
- Idling units in the woods are patrolling their starting location. This prevents them from being picked for an `AttackWave`
- A gunboat is spawned, once for each GunGroup (North & Sout of the river) beeing cleared.
In the original, a gunboat was spawned on entering TriggerCells next to the North Guns, and two times for killing the first two guns. I want to wait for comments on this before changing it. Also I didn't know how to translate eg. cell 2213 from the ini into a CPos.

Import log:

```
-Details----------------------------------------------------
Name   : GDI Mission 9, Hungary
Mission: scg09ea.ini
Source : CNC Gold
Tool   : OpenRA.Utility.exe, bleed 19.07.2019
Reference: https://cnc.fandom.com/wiki/Secure_the_Danube
-Result-----------------------------------------------------
Ignoring unknown MaxBuilding=150 for player BadGuy
Ignoring unknown MaxUnit=150 for player BadGuy
Ignoring unknown Credits=30 for player BadGuy
Ignoring unknown Edge=North for player BadGuy
Ignoring unknown MaxBuilding=150 for player Neutral
Ignoring unknown MaxUnit=150 for player Neutral
Ignoring unknown Edge=North for player Neutral
Ignoring unknown Credits=0 for player Neutral
Ignoring unknown MaxBuilding=150 for player GoodGuy
Ignoring unknown MaxUnit=341 for player GoodGuy
Ignoring unknown Credits=40 for player GoodGuy
Ignoring unknown Edge=East for player GoodGuy
scg09ea.oramap saved.
```

Possible TiberiumTreeCreation adjustment after 16827 is merged.

---
Updates:
The following corrections were made:

- Updated map.png (320 -> 360px width, visually not changed ingame)
- Bridge made undestroyable
- Airstrike SquadSize reduced to 2
- SamSite spawn e1 on death
- Gunboat health reduction. Obelisk will kill the gunboat with one-hit (similar to vanilla)

### CODE REVIEW

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+	Nod.Cash = StartingCash
+	GuardBase()
+end
+
+Trigger.OnAllKilledOrCaptured(NodBase, function()
+	Utils.Do(GDI.GetGroundAttackers(), IdleHunt)
+end)
+
+rigger.OnKilled(nodrefinery, function(building)
```
Comment (abcdefg30) [inline mods/cnc/maps/gdi09/gdi09-AI.lua]: This probably want to be `Trigger`?

Comment (fruestueck) [reply mods/cnc/maps/gdi09/gdi09-AI.lua]: fixed, thanks.
~~I was also thinking about replacing those triggers with:~~ will be rendered obsulet with other PRs anyway

```
...
BaseNuke5 = { type = "nuke", pos = CPos.New(12, 30), cost = 500 }
InfantryProduction = { type = "hand", pos = CPos.New(15, 24), cost = 500 }
VehicleProduction = { type = "afld", pos = CPos.New(3, 27), cost = 2000 }
RebuildBaseOrder = { 
	{ actor = nodrefinery, replacement = BaseRefinery },
	{ actor = nodpower1, replacement = BaseNuke1 },
	{ actor = nodpower2, replacement = BaseNuke2 },
	{ actor = nodpower3, replacement = BaseNuke3 },
	{ actor = nodpower4, replacement = BaseNuke4 },
	{ actor = nodpower5, replacement = BaseNuke5 },
	{ actor = handofnod, replacement = InfantryProduction },
	{ actor = nodairfield, replacement = VehicleProduction },
}
...
RebuildBase = function()
	for k, order in pairs(RebuildBaseOrder) do
		Trigger.OnKilled(order.actor, function()
			BuildBuilding(order.replacement, NodCYard)
		end)
	end
	RebuildBaseOrder = nil
end
```
It may be a bit more clean but I still don't love it.

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+	BuildBuilding(BaseRefinery, NodCYard)
+end)
+
+Trigger.OnKilled(nodpower1, function(building)
```
Comment (fruestueck) [inline mods/cnc/maps/gdi09/gdi09-AI.lua:178]: I guess we can also scratch that building variable in all scripts. But that would be a different PR
Trigger.OnKilled(nodpower1, function(~~building~~)

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 3]
```diff
@@ -0,0 +1,201 @@
+--[[
+   Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
```
Comment (abcdefg30) [inline mods/cnc/maps/gdi09/gdi09-AI.lua]: Just that I don't forget this later on: 2020.

File: mods/cnc/maps/gdi09/gdi09.lua [hunk 4]
```diff
@@ -0,0 +1,163 @@
+--[[
+   Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
```
Comment (abcdefg30) [inline mods/cnc/maps/gdi09/gdi09.lua]: Same here.

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+BaseRefinery = { type = "proc", pos = CPos.New(12, 25), cost = 1500 }
+BaseNuke1 = { type = "nuke", pos = CPos.New( 5, 24), cost = 500 }
+BaseNuke2 = { type = "nuke", pos = CPos.New( 3, 24), cost = 500 }
+BaseNuke3 = { type = "nuke", pos = CPos.New(16, 30), cost = 500 }
+BaseNuke4 = { type = "nuke", pos = CPos.New(14, 30), cost = 500 }
+BaseNuke5 = { type = "nuke", pos = CPos.New(12, 30), cost = 500 }
+InfantryProduction = { type = "hand", pos = CPos.New(15, 24), cost = 500 }
+VehicleProduction = { type = "afld", pos = CPos.New(3, 27), cost = 2000 }
+
+NodGuards = {Actor168, Actor169, Actor170, Actor171, Actor172, Actor181, Actor177, Actor188, Actor189, Actor190 }
```
Comment (abcdefg30) [inline mods/cnc/maps/gdi09/gdi09-AI.lua]: Style nit: Missing space after `{`.

File: mods/cnc/maps/gdi09/gdi09.lua [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	
+	Trigger.AfterDelay(DateTime.Minutes(1), function() SendWaves(1, AutoAttackWaves) end)
+	Trigger.AfterDelay(DateTime.Minutes(3), function() ProduceInfantry(handofnod) end)
+	Trigger.AfterDelay(DateTime.Minutes(3), function() ProduceVehicle(nodairfield) end)
+	
+	local initialArrivingUnits = { Actor175, Actor191, Actor192, Actor193, Actor194, Actor195, Actor196, Actor197, Actor198 }
+	Utils.Do(initialArrivingUnits, function(unit)
+		unit.Move(unit.Location + CVec.New(0, 1), 0)
+	end)
+
```
Comment (abcdefg30) [inline mods/cnc/maps/gdi09/gdi09.lua]: Style nit: Superfluous new line.

File: mods/cnc/maps/gdi09/gdi09.lua [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+		SendGDIReinforcementChinook()
+		Trigger.AfterDelay(DateTime.Seconds(1), SpawnGunboat)
+	end)
+	
+	SecureArea = GDI.AddObjective("Destroy the Nod strike force.")
+	KillGDI = Nod.AddObjective("Kill all enemies!")
+
+	Trigger.AfterDelay(DateTime.Seconds(5), SendGDIReinforcements)
+	
+	AirSupport = GDI.AddSecondaryObjective("Destroy the SAM sites to receive air support.")
```
Comment (abcdefg30) [inline mods/cnc/maps/gdi09/gdi09.lua]: Please use `AddObjective(<text>, "Secondary", false)` instead.

File: mods/cnc/maps/gdi09/gdi09.lua [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+		Trigger.AfterDelay(DateTime.Seconds(team.delay), function() SendWaves(counter + 1, Waves) end)
+	end
+end
+
+StartPatrols = function()
+	for k, team in pairs(Patrols) do
+		local group = 1
+		for type, amount in pairs(team.units) do
+			for i=1,amount,1 do
```
Comment (abcdefg30) [inline mods/cnc/maps/gdi09/gdi09.lua]: Style nit: `for i = 1, amount do`

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+PatrolProductionQueue = { }
+InfantryAttackGroup = { }
+InfantryGroupSize = 5
+InfantryProductionCooldown = DateTime.Minutes(3)
+InfantryProductionTypes = { "e1", "e1", "e1", "e3", "e3", "e4" }
+HarvesterProductionType = { "harv" }
+VehicleAttackGroup = { }
+VehicleGroupSize = 5
+VehicleProductionCooldown = DateTime.Minutes(3)
+ehicleProductionTypes = { "bggy", "bggy", "bggy", "ltnk", "ltnk", "arty" }
```
Comment (abcdefg30) [inline mods/cnc/maps/gdi09/gdi09-AI.lua]: Typo: Missing V.

File: mods/cnc/maps/gdi09/gdi09.lua [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+	Reinforcements.ReinforceWithTransport(GDI, InsertionHelicopterType, GDIHeliReinfUnits, { GDIHeliEntryNorth.Location, GDIHeliLZ.Location }, { GDIHeliLZ.Location + CVec.New(20, 0) })
+end
+
+SendGDIReinforcementChinook = function()
+	Reinforcements.ReinforceWithTransport(GDI, 'tran', nil, { GDIHeliEntryNorth.Location, GDIHeliLZ.Location })
+end
+
+SpawnGunboat = function()
+	Media.PlaySpeechNotification(GDI, "Reinforce")
+	Actor.Create("boat", true, { Owner = GDI, Facing = 0, Location = waypoint23.Location })
```
Comment (abcdefg30) [inline mods/cnc/maps/gdi09/gdi09.lua]: This appears to spawn the boat in the middle of the map and not at the map edge?

File: mods/cnc/maps/gdi09/rules.yaml [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+MLRS:
+	Buildable:
+		Prerequisites: ~disabled
+
+SAM:
+	Buildable:
+		Prerequisites: ~disabled
+	SpawnActorOnDeath:
+		Actor: e1
```
Comment (abcdefg30) [inline mods/cnc/maps/gdi09/rules.yaml]: What is the idea here?

**Review by abcdefg30 — COMMENTED** [review]
Two more things I noticed:
- ~~Do we really want to spawn two boats?~~ Nvm, reading the OP again really helps.
- The boats do not reveal shroud

File: mods/cnc/maps/gdi09/gdi09.lua (hunk 10 above)
Comment (fruestueck) [reply mods/cnc/maps/gdi09/gdi09.lua]: True, back when I ported that mission I did't want to move a waypoint. I'll replace this with CPos.New(62,37)

File: mods/cnc/maps/gdi09/rules.yaml (hunk 11 above)
Comment (fruestueck) [reply mods/cnc/maps/gdi09/rules.yaml]: In the original there was a chance for a sam site to spawn an e1 on destruction, if I remeber correctly. Should we scratch this here?

File: mods/cnc/maps/gdi09/gdi09.lua (hunk 7 above)
Comment (fruestueck) [reply mods/cnc/maps/gdi09/gdi09.lua]: Thanks, didn't know how to convert these to the new style.

File: mods/cnc/maps/gdi09/gdi09-AI.lua (hunk 9 above)
Comment (fruestueck) [reply mods/cnc/maps/gdi09/gdi09-AI.lua]: And I thought I'd handle that already. Nice catch!

**Review by fruestueck — COMMENTED** [review]
Should we leave boats like this (note in op)?

**Comment by fruestueck** [comment]
>     * The boats do not reveal shroud

Tested it, and they do. At least in the CNC Gold version of this mission.

File: mods/cnc/maps/gdi09/rules.yaml (hunk 11 above)
Comment (abcdefg30) [reply mods/cnc/maps/gdi09/rules.yaml]: Imho yes, we don't spawn actors out of any of the other buildings as well (original RA did this for all buildings, so I guess original TD could have been the same).

**Comment by abcdefg30** [comment]
I mean, they didn't reveal shroud when I tested this PR.

**Comment by abcdefg30** [comment]
grafik

**Comment by abcdefg30** [comment]
Hm, if you destroy the gun turrets before attacking the base all ships die to the rocket soldiers at the base entrance, but you can still complete the mission. Maybe we should think about adding an objective here, or allowing the ships to pass in some way. (It doesn't make much sense to me to have a "clear a path for the convoy" objective, but not care if the convoy actually gets through.)

**Comment by fruestueck** [comment]
>boats do not reveal shroud

Strange, it worked before. Boat in gdi1 doesn't reveal shroud eather. I am not sure if this is intended, could be a regression?

> Hm, if you destroy the gun turrets before attacking the base all ships die to the rocket soldiers at the base entrance, but you can still complete the mission. Maybe we should think about adding an objective here, or allowing the ships to pass in some way. (It doesn't make much sense to me to have a "clear a path for the convoy" objective, but not care if the convoy actually gets through.)

I agree that it doesn't make any sense. It kinda felt pointless to care for the objective while testing it in the original. Back then I tried to port the missions with low modifications to the logic.
As far as I understand, it's intendet that the ships get wrecked by the obelisk to show the importance of getting rid of that base.

But I am not sure how to alter the goals/logic to make more sense of it without altering to much.
I guess, if a player clears the base before sending a ship and that ship makes it through, the mission could be accomplished or at least remove the ship from the west border and print some kind of message "reinforcement made it throug, but still clear the base to allow further boats...".

Regarding the rockets: Initially they did't kill the ship, and it took the obelisk several hits as well, so I've lowered the HP of the boat maby a bit too much. As the rockets have a bigger range than in the original, this changes things a bit. I could increase the boats health again, as well as the obelisks damage output in order to keep the obelisk one-hitting it. Then the rockets would be as unimportant as in the original.

**Comment by fruestueck** [comment]
> The boats do not reveal shroud

Oh, that was different back when I made this. The boat of mission gdi1 doesn't eather. Is this intended or a regression?

> regarding the objectives and the boat

I agree that the objective feels kind of pointles when the ships have no effect but I am not sure if and how to alter the mission. We can't let the player win if one ship makes it through as the mission states to raise the base. But we could remove ships that made it to the other side and print some message.
Looking back on the [mission details](https://cnc.fandom.com/wiki/Secure_the_Danube) it seems that this was the first mission revealing an obelisk and the ships are simply there to show its power, i guess.

The rockets shouldn't kill a boat. I've lowered its hp earlyer to give the obelisk the power of one-hitting it (as in the original). I can increase the boats hp and the obelisks dmg to keep this behaviour and render the rockets ineffective.

**Comment by abcdefg30** [comment]
> I can increase the boats hp and the obelisks dmg to keep this behaviour and render the rockets ineffective.

A better idea than changing the damage values might be assigning a new armor type to the boats and make obelisks strong against that armor but rockets weak. That way we don't affect how the obelisk does against other units.

> But we could remove ships that made it to the other side and print some message.

Good idea, we should do that.
We can focus on altering the mission to make more sense in a follow-up PR if you want, that way we don't block progress here over a detail.

> Is this intended or a regression?

Looks like a regression. I filed #17585 to fix it.

**Comment by fruestueck** [comment]
> A better idea than changing the damage values might be assigning a new armor type to the boats and make obelisks strong against that armor but rockets weak. That way we don't affect how the obelisk does against other units.

Could you point to an example implementation of custom armor for a mission. I've searched through the yamls with [https://github.com/OpenRA/OpenRA/search?l=YAML&q=armor](OpenRA/search?l=YAML&q=armor) without any luck.

> We can focus on altering the mission to make more sense in a follow-up PR if you want, that way we don't block progress here over a detail.

Agreed. A futur PR could also add difficulties. For now, I've added a Trigger.OnEnteredFootprint that disposes the boat and prints a line. I've no idea why it isn't working. Could you have a look on it? Do celltriggers maby not work in cnc waters?

> Looks like a regression. I filed #17585 to fix it.

Works like a charm : )

File: mods/cnc/maps/gdi09/gdi09.lua [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+	Trigger.OnAllKilled(NodBunkersNorth, function()
+		GDI.MarkCompletedObjective(DestroyBunkers)
+		Trigger.AfterDelay(DateTime.Seconds(1), SpawnGunboat)
+	end)
+	Trigger.OnAllKilled(NodBunkersSouth, function()
+		GDI.MarkCompletedObjective(DestroyBunkers)
+		SendGDIReinforcementChinook()
+		Trigger.AfterDelay(DateTime.Seconds(1), SpawnGunboat)
+	end)
+	Trigger.OnEnteredFootprint({ CPos.New(60,37) }, function(a, id)
```
Comment (fruestueck) [inline mods/cnc/maps/gdi09/gdi09.lua]: Should be `CPos.New(2,37)` instead of `(60,37)` (for testing) but it isn't working right now.

Comment (Mailaender) [reply mods/cnc/maps/gdi09/gdi09.lua]: To enhance readability you could move that to a variable:
```lua
BoatTrigger = { CPos.New(2,37) }
[...]
Trigger.OnEnteredFootprint(BoatTrigger, function(a, id)
```

File: mods/cnc/maps/gdi09/gdi09.lua [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+		if a.type == "boat" then
```
Comment (Mailaender) [inline mods/cnc/maps/gdi09/gdi09.lua]: Try changing it to
```suggestion
		if a.Type == "boat" then
```

File: mods/cnc/maps/gdi09/gdi09.lua (hunk 12 above)
Comment (fruestueck) [reply mods/cnc/maps/gdi09/gdi09.lua]: Thanks!

File: mods/cnc/maps/gdi09/gdi09.lua (hunk 13 above)
Comment (fruestueck) [reply mods/cnc/maps/gdi09/gdi09.lua]: I've changed it and tested a bit more.

A simple message  for debugging purpose
`Media.DisplayMessage("Actor " .. a.Type .. " entered TriggerFootprint")` 
before the `if a.Type == "boat"` check revealed that an LST does trigger OnEnteredFootprint while the gunboat doesn't. Is this a possible bug or a misconfig of my boat?

Comment (Mailaender) [reply mods/cnc/maps/gdi09/gdi09.lua]: The gunboat is a bit of a special case with the `TDGunboat` trait which does all the movement hacks. Maybe it doesn't trigger properly.

**Comment by fruestueck** [comment]
> A better idea than changing the damage values might be assigning a new armor type to the boats and make obelisks strong against that armor but rockets weak. That way we don't affect how the obelisk does against other units.

Regarding the Armor and Obelisk behaviour: I've opted against an introduction of a new armor type as I've no Idea how and instead removed the boat health boost. That way we don't overcomplicate things and the balancing feels also right (at least to me). The rocket soldiers won't affect the gunboats too much, and the obelisk still gets rid of a boat in no time with two-shots.
Finally, the gunturrets still pose a threat to the gunboat wich seems allright as they are "the showstopper" for the boats in this mission by missiondesign anyway.

Regarding the disposal of the gunboats when reaching the west border of the map, we have two options:

1. Check every n ticks if the boats position matches the west edge of the river which seems a dirty hack OR
2. Leave it for now like it is. This would mean that the boats won't get removed (which would match the vanilla gameplay) and as soon as #17780 gets implemented the already added code would dispose the boats and display a message.

I'm more in favor of option 2. Possibly without the Trigger in advance. : )

**Review by Mailaender — DISMISSED** [review]

**Comment by pchote** [comment]
Needs a rebase now!

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+   Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+   This file is part of OpenRA, which is free software. It is made
+   available to you under the terms of the GNU General Public License
+   as published by the Free Software Foundation, either version 3 of
+   the License, or (at your option) any later version. For more
+   information, see COPYING.
+]]
+
+AttackPaths = { { waypoint7 }, { waypoint8 } }
+NodBase = { handofnod, nodairfield, nodrefinery, NodCYard, nodpower1, nodpower2, nodpower3, nodpower4, nodpower5, gun5, gun6, gun7, gun8, nodsilo1, nodsilo2, nodsilo3, nodsilo4, nodobelisk}
```
Comment (pchote) [inline mods/cnc/maps/gdi09/gdi09-AI.lua]: Missing space before the final `}`

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+BaseNuke5 = { type = "nuke", pos = CPos.New(12, 30), cost = 500 }
+InfantryProduction = { type = "hand", pos = CPos.New(15, 24), cost = 500 }
+VehicleProduction = { type = "afld", pos = CPos.New(3, 27), cost = 2000 }
+
+NodGuards = { Actor168, Actor169, Actor170, Actor171, Actor172, Actor181, Actor177, Actor188, Actor189, Actor190 }
+
+BaseBuildings = { BaseRefinery, BaseNuke1, BaseNuke2, BaseNuke3, BaseNuke4, InfantryProduction, VehicleProduction }
+
+BuildBuilding = function(building, cyard)
+	if CyardIsBuilding or Nod.Cash < building.cost then
```
Comment (pchote) [inline mods/cnc/maps/gdi09/gdi09-AI.lua]: `CyardIsBuilding` appears to be used here before it is first assigned. Do we need a `CyardIsBuilding = false` in the definitions at the top of the file?

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+		Trigger.AfterDelay(DateTime.Seconds(10), function() BuildBuilding(building, cyard) end)
+		return
+	end
+
+	CyardIsBuilding = true
+
+	Nod.Cash = Nod.Cash - building.cost
+	Trigger.AfterDelay(Actor.BuildTime(building.type), function()
```
Comment (pchote) [inline mods/cnc/maps/gdi09/gdi09-AI.lua:48]: We really should define a way for missions to produce/place structures properly!

This overlaps strongly with my comment in https://github.com/OpenRA/OpenRA/pull/17129#discussion_r365551639 - I don't expect (or want) anything to do be done about it in this PR!

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+InfantryProductionTypes = { "e1", "e1", "e1", "e3", "e3", "e4" }
+HarvesterProductionType = { "harv" }
+VehicleAttackGroup = { }
+VehicleGroupSize = 5
+VehicleProductionCooldown = DateTime.Minutes(3)
+VehicleProductionTypes = { "bggy", "bggy", "bggy", "ltnk", "ltnk", "arty" }
+StartingCash = 14000
+
+BaseRefinery = { type = "proc", pos = CPos.New(12, 25), cost = 1500 }
+BaseNuke1 = { type = "nuke", pos = CPos.New( 5, 24), cost = 500 }
```
Comment (pchote) [inline mods/cnc/maps/gdi09/gdi09-AI.lua]: stray space before the 5.

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+InfantryProductionCooldown = DateTime.Minutes(3)
+InfantryProductionTypes = { "e1", "e1", "e1", "e3", "e3", "e4" }
+HarvesterProductionType = { "harv" }
+VehicleAttackGroup = { }
+VehicleGroupSize = 5
+VehicleProductionCooldown = DateTime.Minutes(3)
+VehicleProductionTypes = { "bggy", "bggy", "bggy", "ltnk", "ltnk", "arty" }
+StartingCash = 14000
+
+BaseRefinery = { type = "proc", pos = CPos.New(12, 25), cost = 1500 }
```
Comment (pchote) [inline mods/cnc/maps/gdi09/gdi09-AI.lua]: Can we use `Actor.Cost("proc")` instead of hardcoding costs that may change?

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+
+		Trigger.OnKilled(actor, function()
+			BuildBuilding(building, cyard)
+		end)
+
+		RepairBuilding(Nod, actor, 0.75)
+	end)
+end
+
+CheckForHarvester = function()
```
Comment (pchote) [inline mods/cnc/maps/gdi09/gdi09-AI.lua]: This method name doesn't tell us what it does. Can we please call this something like `HasHarvester`?

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+	local harv = Nod.GetActorsByType("harv")
+	return #harv > 0
+end
+
+GuardBase = function()
+	Utils.Do(NodBase, function(building)
+		Trigger.OnDamaged(building, function()
+			Utils.Do(NodGuards, function(guard)
+				if not guard.IsDead and not building.IsDead then
```
Comment (pchote) [inline mods/cnc/maps/gdi09/gdi09-AI.lua]: the `building.IsDead` check can move outside this inner `Utils.Do`

File: mods/cnc/maps/gdi09/gdi09-AI.lua [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+
+StartAI = function()
+	RepairNamedActors(Nod, 0.75)
+
+	Nod.Cash = StartingCash
+	GuardBase()
+end
+
+Trigger.OnAllKilledOrCaptured(NodBase, function()
+	Utils.Do(GDI.GetGroundAttackers(), IdleHunt)
```
Comment (pchote) [inline mods/cnc/maps/gdi09/gdi09-AI.lua]: Am I reading this correctly as the script forcing all of the *player*'s units to hunt? Why?

File: mods/cnc/maps/gdi09/gdi09.lua [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+
+Patrols = 
+{ 
+	grd1 = { units = { ['e3'] = 3 }, waypoints = Patrol1Waypoints, wait = 40, initialWaypointPlacement = { 1 } },
+	grd2 = { units = { ['e1'] = 2, ['e3'] = 2, ['e4'] = 2 }, waypoints = Patrol2Waypoints, wait = 20, initialWaypointPlacement = { 4, 10, 1 } }
+}
+
+AutoAttackWaves = { Nod1, Nod2, Nod3, Nod4, Nod5, Auto1, Auto2, Auto3, Auto4, Auto5, Auto6, Auto7, Auto8 }
+
+StationaryGuards = { Actor174, Actor173, Actor182, Actor183, Actor184, Actor185, Actor186, Actor187 , Actor199, Actor200, Actor201, Actor202, Actor203, Actor204}
```
Comment (pchote) [inline mods/cnc/maps/gdi09/gdi09.lua]: Missing space before the `}`

File: mods/cnc/maps/gdi09/gdi09-AI.lua (hunk 15 above)
Comment (abcdefg30) [reply mods/cnc/maps/gdi09/gdi09-AI.lua]: Not in Lua.

File: mods/cnc/maps/gdi09/gdi09-AI.lua (hunk 16 above)
Comment (abcdefg30) [reply mods/cnc/maps/gdi09/gdi09-AI.lua:48]: By now all of our TD missions (much to my own displeasure) use this system.

File: mods/cnc/maps/gdi09/gdi09-AI.lua (hunk 18 above)
Comment (abcdefg30) [reply mods/cnc/maps/gdi09/gdi09-AI.lua]: Good idea, but Imho we shouldn't waste much more time on this system anyway.

File: mods/cnc/maps/gdi09/gdi09-AI.lua (hunk 21 above)
Comment (abcdefg30) [reply mods/cnc/maps/gdi09/gdi09-AI.lua]: That is standard when all base buildings are destroyed, so that you don't have to hunt every unit down. Didn't the original do the same?

Comment (pchote) [reply mods/cnc/maps/gdi09/gdi09-AI.lua]: The original had all enemy units attack the player, IIRC.

Comment (abcdefg30) [reply mods/cnc/maps/gdi09/gdi09-AI.lua]: Ah my bad, I see that you meant `GDI` and `Nod` are mixed up here.

File: mods/cnc/maps/gdi09/gdi09-AI.lua (hunk 18 above)
Comment (fruestueck) [reply mods/cnc/maps/gdi09/gdi09-AI.lua]: Thanks for the hint. Could be changed in the other missions as well but I guess it will be replaced soon ayway?

File: mods/cnc/maps/gdi09/gdi09-AI.lua (hunk 21 above)
Comment (fruestueck) [reply mods/cnc/maps/gdi09/gdi09-AI.lua]: Whops, fixed.

File: mods/cnc/maps/gdi09/gdi09.lua (hunk 22 above)
Comment (fruestueck) [reply mods/cnc/maps/gdi09/gdi09.lua]: How do you spot these space mistakes? ( :

**Comment by fruestueck** [comment]
Applied changes and rebased.

**Comment by pchote** [comment]
Needs a rebase again sorry

**Comment by fruestueck** [comment]
Rebased.

**Comment by fruestueck** [comment]
Fixed OpenRA.sln (removed lua scripts) rebase.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/1cf6a0bdbb62ee0665a3afe3fca1c353ef6a0872...ed16f89b78a83235cbe337e26e1488da12960869

70e1c4b
